cpp/src/arrow/compute/exec/plan_test.cc

```diff
@@ -374,6 +374,108 @@ TEST(ExecPlanExecution, SourceSinkError) {
               Finishes(Raises(StatusCode::Invalid, HasSubstr("Artificial"))));
 }
 
+TEST(ExecPlanExecution, SourceConsumingSink) {
+  for (bool slow : {false, true}) {
+    SCOPED_TRACE(slow ? "slowed" : "unslowed");
+
+    for (bool parallel : {false, true}) {
+      SCOPED_TRACE(parallel ? "parallel" : "single threaded");
+      ASSERT_OK_AND_ASSIGN(auto plan, ExecPlan::Make());
+      uint32_t batches_seen = 0;
+      Future<> finish = Future<>::Make();
+      struct TestConsumer : public SinkNodeConsumer {
+        TestConsumer(uint32_t* batches_seen, Future<> finish)
```

Comment:
In the parallel case, shouldn't this be an atomic?

Reply:
(I suppose with only 2 batches, it's unlikely to run into this, but…)

Reply:
Done.